A pass on the documentation of records (#15489)
coq/coq: doc/sphinx/language/core/records.rst

         Module B. Record S := { f : nat }. End B.

      .. coqtop:: all

         Check {| A.f := 0 |}.
         Check {| B.f := 0 |}.

.. _record_as_clause:

   .. example:: Using the "as" clause in a record definition

      .. coqtop:: all

         Record MyRecord := { myfield : nat } as VarName.
         About myfield. (* observe the MyRecord variable is named "VarName" *)
         
         (* make "VarName" implicit without having to rename the variable,
            which would be necessary without the "as" clause *)
         Arguments myfield {VarName}.   (* make "VarName" an implicit parameter *)
         Check myfield.
         Check (myfield (VarName:={| myfield := 0 |})).

   .. exn:: Records declared with the keyword Record or Structure cannot be recursive.

      The record name :token:`ident` appears in the type of its fields, but uses
      the :cmd:`Record` command. Use  the :cmd:`Inductive` or
      :cmd:`CoInductive` command instead.

   .. exn:: @ident already exists

      The fieldname :n:`@ident` is already defined as a global.

   .. warn:: @ident__1 cannot be defined because the projection @ident__2 was not defined

>       :undocumented:

Reviewer: Can you reintroduce the documentation for this warning, i.e.: ```The type of the projections :token:`ident` depends on previous projections which themselves could not be defined.```?

Reviewer 2: Got it.  Updated.

   the types in the block are records.  These commands automatically generate
   induction schemes.  Enable the :flag:`Nonrecursive Elimination Schemes` flag
   to enable automatic generation of elimination schemes for :cmd:`Record`.
   See :ref:`proofschemes-induction-principles`.

   :n:`{? > }`
     If specified, the constructor is declared as
     a coercion from the class of the last field type to the record name
     (this may fail if the uniform inheritance condition is not
     satisfied). See :ref:`coercions`.

   :n:`@ident_decl`
     The :n:`@ident` within is the record name.

   :n:`{* @binder }`
     :n:`@binder`\s may be used to declare the
     :term:`inductive parameters <inductive parameter>` of the record.

   :n:`: @sort`
     The sort the record belongs to.  The default is :n:`Type`.

   :n:`:= {? @ident }`
     :n:`@ident` is the name of the record constructor.  If omitted,
     the name defaults to :n:`Build_@ident` where :n:`@ident` is the record name.

   :n:`as {? @ident}`
     Specifies the name used to refer to the argument corresponding to the
     record in the type of projections.  If not specified, the name is the first

>      letter of the record name.  See :ref:`example <record_as_clause>`.

Author: We could eventully add that is the lowercase form of the first letter of the type.

We could also add that in the case of a `Class`, it is by default the full name of the class. (This does seem a bad default choice actually, and, retrospectively, we could have done that for records also. Maybe too late to change and the `as` clause is now the way to have control on it.)

Reviewer 2: I added a few words on `Class` with an example.  Perhaps it should be under the `Class` command description, but I think it's OK here for now.  (I think of `Record` as the primary command here and that the others are specializations (and thus secondary)).

One thing I don't understand is why there are so many ways to define records, i.e. `Record`, `Inductive`, `CoInductive`, `Class` and `Instance`.  I've not tried to understand typeclasses, so the need for the last two is a bit mysterious.  If one of you can provide a succinct way to distinguish those two, I'll add it in the text for the `Record` command where I mention `Inductive` and `CoInductive`.

Author: `Instance` inhabits a `Class`, so it is not a type but instead something equivalent to `{| .... |}`. `Class`es are `Record`s with extra information for automatic inference of inhabitants of the record. (Actually, it is a bit more subtle in that there are classes which are just definitions, that is, when a class is a record with only one field, we skip the indirection and treat the class as equal to the contents of the field; so technically `Class` is either a `Record` or a `Definition`. Again, when a Class is a Definition, it is to indicate extra information for automatic inference of the inhabitants).

When a `Record` is recursive, one need to know if it is `Inductive`, that is only with a finite nesting of the constructors (or more generally with only a well-founded nesting of the constructor), or `CoInductive`, that is with possibly infinite inhabitants (the prototypical example being streams: `CoInductive stream A := {hd:A; tl: stream A}.`).

When a record is not recursive, we don't need to tell if it is Inductive or CoInductive, so we use the word `Record` instead, to avoid having to arbitrarily choose an information Inductive/CoInductive which would be pointless.

Author: > I added a few words on Class with an example. Perhaps it should be under the Class command description, but I think it's OK here for now. (I think of Record as the primary command here and that the others are specializations (and thus secondary)).

I think that is ok as it is. This question of naming the inhabitant of the record or class is anyway relatively marginal compared to the logical aspects of a record.

Reviewer 2: I'll just say this for now:

The :cmd:`Class` command can be used to define records with extra information that permits Coq to automatically infer the inhabitants of the record.

   In :n:`@record_field`:

     :n:`@attribute`, if specified, can only be :attr:`canonical`.

     :n:`@name` is the field name.  Since field names define projections, you can't
     reuse the same field name in two different records in the same module.  This
     :ref:`example <reuse_field_name>` shows how to reuse the same field
     name in multiple records.

     :n:`@field_spec` can be omitted only when the type of the field can be inferred
     from other fields. For example: the type of :n:`n` can be inferred from
     :n:`npos` in :n:`Record positive := { n; npos : 0 < n }`.

     :n:`| @natural`
       Specifies the priority of the field.  It is only allowed in :cmd:`Class` commands.

   In :n:`@field_spec`:

     - :n:`{+ @binder } : @of_type` is equivalent to
       :n:`: forall {+ @binder } , @of_type`

     - :n:`{+ @binder } := @term` is equivalent to

>        :n:`:= fun {* @binder } => @term`

Author: ```suggestion
       :n:`:= fun {+ @binder } => @term`
```

     - :n:`{+ @binder } @of_type := @term` is equivalent to

>        :n:`: forall {* @binder } , @type := fun {* @binder } => @term`

Author: ```suggestion
       :n:`: forall {+ @binder } , @type := fun {* @binder } => @term`
```

         Module B. Record S := { f : nat }. End B.

      .. coqtop:: all

         Check {| A.f := 0 |}.
         Check {| B.f := 0 |}.

.. _record_as_clause:

   .. example:: Using the "as" clause in a record definition

      .. coqtop:: all

         Record MyRecord := { myfield : nat } as VarName.
         About myfield. (* observe the MyRecord variable is named "VarName" *)

         (* make "VarName" implicit without having to rename the variable,
            which would be necessary without the "as" clause *)
         Arguments myfield {VarName}.   (* make "VarName" an implicit parameter *)
         Check myfield.
         Check (myfield (VarName:={| myfield := 0 |})).

   .. exn:: Records declared with the keyword Record or Structure cannot be recursive.

      The record name :token:`ident` appears in the type of its fields, but uses
      the :cmd:`Record` command. Use  the :cmd:`Inductive` or
      :cmd:`CoInductive` command instead.

   .. exn:: @ident already exists

      The fieldname :n:`@ident` is already defined as a global.

   .. warn:: @ident__1 cannot be defined because the projection @ident__2 was not defined

>       The type of the projection :token:`ident` depends on previous projections which

Author: Add __1 to the name? (But I vaguely remember this would not work, maybe.)

Reviewer: It works if we make it
```suggestion
      The type of the projection :n:`@ident__1` depends on previous projections which
```

      themselves could not be defined.

   .. warn:: @ident cannot be defined.

      The projection cannot be defined.  This message is followed by an explanation
      of why it's not possible, such as:

      #. The :term:`body` of :token:`ident` uses an incorrect elimination for
         :token:`ident` (see :cmd:`Fixpoint` and :ref:`Destructors`).

   .. warn:: @ident__field cannot be defined because it is informative and @ident__record is not

      The projection for the named field can't be defined.
      For example, :n:`Record foo:Prop := { x:Type }` generates the message
      "x cannot be defined ... and foo is not".  Proofs (objects of sort :n:`Prop`)
      are supposed to be non-distinguishable.  If you have two inhabitants of
      :n:`Type`, such as :n:`%{%| x := nat %|%}` and :n:`%{%| x := bool %|%}`, they are

>       distinguishable (i.e. informative) and are therefore prohibited.

Author: An example using simpler objects (i.e. not a type of types) would be to simply take `Record foo:Prop := { x:nat }`, then continuing e.g. as follows: ``Inhabitants of :n:`nat` are distinguishable (i.e. informative), so to preserve say :n:`%{%| x := 0 %|%}` and :n:`%{%| x := 1 %|%}` from being distinguished, the projection has to be prohibited.``

We could even take `bool` instead of `nat` which  would be the smallest counterexample to leaking information via projections.

Reviewer 2: How about this?

![image](https://user-images.githubusercontent.com/1253341/154819565-89510f86-7c7e-4667-90dd-cfa6cd78c01f.png)

(OTOH, I don't quite see how a record could be a proof.  No need to explain now.)

Author: I think that's ok (records -> record?)

> (OTOH, I don't quite see how a record could be a proof. No need to explain now.)

A typical example of proposition which it makes sense to defined as a `Record` is `Record and (A B:Prop) := {proj1:A ; proj2:B}.` (even if it is not done like this in the library).

         Admitted.

         (* Record form: top and bottom can be inferred from other fields *)
         Definition half :=
           {| negative := false;
              Rat_bottom_nonzero := O_S 1;
              Rat_irreducible := one_two_irred |}.

         (* Application form: use the constructor and provide values for all the fields
            in order.  "mkRat" is defined by the Record command *)
         Definition half' := mkRat true 1 2 (O_S 1) one_two_irred.

Accessing fields (projections)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

   .. insertprodn term_projection term_projection

   .. prodn::
      term_projection ::= @term0 .( @qualid {? @univ_annot } {* @arg } )
      | @term0 .( @ @qualid {? @univ_annot } {* @term1 } )

   The value of a field can be accessed using *projection form* (:n:`@term_projection`,
   shown here) or with *application form* (see :n:`@term_application`) using the
   projection function associated with the field.  Don't forget the parentheses for the
   projection form.
   Glossing over some syntactic details, the two forms are:

   - :n:`@qualid__record.( {? @ } @qualid__field {* @arg })` (projection) and

>    - :n:`{? @ } @qualid__field {* @arg } @qualid__record` (application)

Author: That's strange. In the sphinx rendering, the space before `(application)` is barely perceptible.

Reviewer 2: Non-breaking space to the rescue:

![image](https://user-images.githubusercontent.com/1253341/154818718-cc9c48ab-6f0c-42a4-a9ee-12dea5bf56ee.png)

Author: Much better indeed.

   :name: Record; Structure

   .. insertprodn record_definition field_spec

   .. prodn::
      record_definition ::= {? > } @ident_decl {* @binder } {? : @sort } {? := {? @ident } %{ {*; @record_field } {? ; } %} {? as @ident } }
      record_field ::= {* #[ {*, @attribute } ] } @name {? @field_spec } {? %| @natural }
      field_spec ::= {* @binder } @of_type
      | {* @binder } := @term
      | {* @binder } @of_type := @term

   Defines a non-recursive record type, creating projections for each field
   that has a name other than `_`. The field body and type can depend on previous
   fields, so the order of fields in the definition may matter.

   Use the :cmd:`Inductive` and :cmd:`CoInductive` commands to define recursive
   (inductive or coinductive) records.  These commands also permit defining
   mutually recursive records provided that all of
   the types in the block are records.  These commands automatically generate
   induction schemes.  Enable the :flag:`Nonrecursive Elimination Schemes` flag
   to enable automatic generation of elimination schemes for :cmd:`Record`.
   See :ref:`proofschemes-induction-principles`.

>    The :cmd:`Class` command can be used to define records with extra
>    information that permits Coq to automatically infer the inhabitants of the record.

Author: I proposed it but I find the "extra information" a bit weak. Maybe:
```suggestion
   The :cmd:`Class` command can be used to define records equipped
   with a mechanism of :ref:`typeclasses` that permits Coq to automatically infer the inhabitants of the record.
```

   mutually inductive or coinductive records, provided all of
   the types in the block are records.  These commands automatically generate
   induction schemes.  Enable the :flag:`Nonrecursive Elimination Schemes` flag
   to enable automatic generation of elimination schemes for :cmd:`Record`.
   See :ref:`proofschemes-induction-principles`.

   :n:`{? > }`
     If specified, the constructor is declared as
     a coercion from the class of the last field type to the record name
     (this may fail if the uniform inheritance condition is not
     satisfied). See :ref:`coercions`.

   :n:`@ident_decl`
     The :n:`@ident` within is the name of the record.

   :n:`{* @binder }`
     :n:`@binder`\s may be used to declare the *parameters* of the record.

   :n:`: @sort`
     The sort the record belongs to.  The default is :n:`Type`.

   :n:`:= {? @ident }`
     :n:`@ident` is the name of the record constructor.  If omitted,
     the name defaults to :n:`Build_@ident` where :n:`@ident` is the record name.

   :n:`as {? @ident}`
     Specifies the name used to refer to the argument corresponding to the
     record in the type of projections.

>      .. todo PR: clarify above description, maybe give an example

Reviewer 2: Can you help me improve the description, maybe give an example?  I don't understand this.  Thanks.

Reviewer 2: OK, now how do I access or see `HIMOM`?  `About HIMOM` doesn't work.  Nor does `Definition xx := {| bar := 3 |}.  Eval compute in xx.(HIMOM).`

Author: `HIMOM` is not a constant. It is a name hint, like other names managed by the `Arguments` command. Typical ways to observe it is as follows:
```
Arguments bar {HIMOM}.
Check  (bar (HIMOM:={|bar:=0|})). 
```

Reviewer 2: Ah, OK.  I added an example, but I don't see how this argument is useful, which would be nice to mention.

Reviewer 2: @herbelin Perhaps you can give me a short explanation of when/why the "as" is useful?  My impression is that it's for convenience.

@Zimmi48 Would you take a look at what we have and assign?  Thanks.

Author: > @herbelin Perhaps you can give me a short explanation of when/why the "as" is useful? My impression is that it's for convenience.

Yes, it is to give user control on the name. Virtually, this could be applied also to the name used in the production of induction schemes, or when naming hypotheses of this type in a goal, etc.

Reviewer 2: Ah, the question I meant to ask is why is it beneficial to have HIMOM in `Check  (bar (HIMOM:={|bar:=0|})).` instead of `Check  (bar {|bar:=0|})`?  I see they report different types but I expected they would be equivalent.  Is it just a shortcut so the reduction in Compute isn't needed?  I'm probably misunderstanding something elementary.

The benefit is just "(* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)"?

```
Record foo := { bar : nat ; } as HIMOM.

About bar.
Compute (bar {| bar := 0 |}).   (* = 0 : nat *)
Check (bar {| bar := 0 |}).     (* bar {| bar := 0 |} : nat *)
Arguments bar {HIMOM}.
Check  (bar (HIMOM:={|bar:=0|})).    (* bar : nat *)
Compute (bar {| bar := 0 |}).   (* error (below) *)
```

![image](https://user-images.githubusercontent.com/1253341/153935774-ac8798c2-364c-489e-89e5-d63131883fb6.png)

Reviewer 3: @jfehrle Well that will error because you've made the argument implicit. This shows why it's important to have the name for the variable, because after that argument is implicit there are only two ways of passing that argument. Using the `bar (HIMOM:=_)` syntax or making all implicit arguments explicit with the `@` syntax: `@bar _`.

   :attr:`universes(template)`, :attr:`universes(cumulative)`,
   :attr:`private(matching)` and :attr:`projections(primitive)` attributes.

   .. example:: Defining a record

      The set of rational numbers may be defined as:

      .. coqtop:: reset all

         Record Rat : Set := mkRat
          { negative : bool
          ; top : nat
          ; bottom : nat
          ; Rat_bottom_nonzero : 0 <> bottom
          ; Rat_irreducible :
              forall x y z:nat, (x * y) = top /\ (x * z) = bottom -> x = 1
          }.

      The :n:`Rat_*` fields depend on :n:`top` and :n:`bottom`.
      :n:`Rat_bottom_nonzero` is a proof that :n:`bottom` (the denominator)
      is not zero.  :n:`Rat_irreducible` is a proof that the fraction is in
      lowest terms.

.. _reuse_field_name:

   .. example:: Reusing a field name in multiple records

      .. coqtop:: in

         Module A. Record R := { f : nat }. End A.
         Module B. Record S := { f : nat }. End B.

      .. coqtop:: all

         Check {| A.f := 0 |}.
         Check {| B.f := 0 |}.

.. _record_as_clause:

   .. example:: Using the "as" clause in a record definition

      .. coqtop:: all

>          Record foo := { bar : nat } as HIMOM.
>          About bar.
>          Arguments bar {HIMOM}.   (* make "foo" an implicit parameter *)
>          Check bar.
>          Check (bar (HIMOM:={|bar:=0|})).

Reviewer 3: The names here aren't great, though they were sufficient for discussion. How about:

```suggestion
         Record MyRecord := { myfield : nat } as VarName.
         About myfield. (* observe the MyRecord variable is named VarName *)
         
         (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)
         Arguments myfield {VarName}.   (* make "VarName" an implicit parameter *)
         Check myfield.
         Check (myfield (VarName:={| myfield := 0 |})).
```

Record types
----------------

> 

Reviewer: Would it make sense to add a subsection here "Defining a record type" (for symmetry with "Constructing records" and "Accessing fields")?

   .. insertprodn record_definition field_spec

   .. prodn::
      record_definition ::= {? > } @ident_decl {* @binder } {? : @sort } {? := {? @ident } %{ {*; @record_field } {? ; } %} {? as @ident } }
      record_field ::= {* #[ {*, @attribute } ] } @name {? @field_spec } {? %| @natural }
      field_spec ::= {* @binder } @of_type
      | {* @binder } := @term
      | {* @binder } @of_type := @term

   Defines a non-recursive record type, creating projections for each field
   that has a name other than `_`. The field body and type can depend on previous
   fields, so the order of fields in the definition may matter.

   Use the :cmd:`Inductive` and :cmd:`CoInductive` commands to define recursive
   (inductive or coinductive) records.  These commands also permit defining
   mutually recursive records provided that all of
   the types in the block are records.  These commands automatically generate
   induction schemes.  Enable the :flag:`Nonrecursive Elimination Schemes` flag
   to enable automatic generation of elimination schemes for :cmd:`Record`.
   See :ref:`proofschemes-induction-principles`.

   :n:`{? > }`
     If specified, the constructor is declared as

>      a coercion from the class of the last field type to the record name

Reviewer 3: I can confirm that I and many users found this concept confusing at first.

>      a coercion from the class of the last field type to the record name

Reviewer 2: If you remind me, I'll do it shortly after this PR is done.

>      a coercion from the class of the last field type to the record name

Reviewer 3: I've created an issue https://github.com/coq/coq/issues/15680

>      a coercion from the class of the last field type to the record name

Reviewer 2: That and a reminder will probably do the trick :-).  Too many little details to remember.

     (this may fail if the uniform inheritance condition is not
     satisfied). See :ref:`coercions`.

   :n:`@ident_decl`
     The :n:`@ident` within is the record name.

   :n:`{* @binder }`

>      :n:`@binder`\s may be used to declare the *parameters* of the record.

Reviewer 2: "parameters" is too general a term.  Perhaps "record parameters".  I thought this wording was a bit vague but still understandable.  Seemed like it might take a while to understand it well enough to explain clearly.

>      :n:`@binder`\s may be used to declare the *parameters* of the record.

Reviewer: My question was precisely whether "record parameters" should be interpreted in the same way as "inductive parameters".

>      :n:`@binder`\s may be used to declare the *parameters* of the record.

Reviewer 2: @herbelin, do you know (and can suggest wording)?

>      :n:`@binder`\s may be used to declare the *parameters* of the record.

Reviewer 3: How about saying `the (inductive) *parameters* of the record`?

>      :n:`@binder`\s may be used to declare the *parameters* of the record.

Reviewer 2: I wouldn't add "inductive" unless I understand that it's different from "record parameters".  (I don't change things I know I don't understand.)

>      :n:`@binder`\s may be used to declare the *parameters* of the record.

Reviewer 3: But to answer @Zimmi48's question, parameters for records are the same for inductive types (which records happen to be a special case of).

>      :n:`@binder`\s may be used to declare the *parameters* of the record.

Reviewer: Thanks, then I reiterate my suggestion as such:
```suggestion
     :n:`@binder`\s may be used to declare the :term:`parameters` of the record.
```

Where "parameters" would be defined in the inductive chapter. But this is just a suggestion, and could also be deferred.

>      :n:`@binder`\s may be used to declare the *parameters* of the record.

Reviewer 2: Got it.  I made it "inductive parameters".

   :n:`: @sort`
     The sort the record belongs to.  The default is :n:`Type`.

   :n:`:= {? @ident }`
     :n:`@ident` is the name of the record constructor.  If omitted,
     the name defaults to :n:`Build_@ident` where :n:`@ident` is the record name.

   :n:`as {? @ident}`
     Specifies the name used to refer to the argument corresponding to the
     record in the type of projections.  If not specified, the name is the first
     letter of the record name.  See :ref:`example <record_as_clause>`.

   In :n:`@record_field`:

     :n:`@attribute`, if specified, can only be :n:`canonical`.

     :n:`@name` is the field name.  Since field names define projections, you can't

>      reuse the same field name in two different records in the same scope.  This

Reviewer 2: Yup.  At some point I want to have a section that describes all the different types of names Coq uses and their scopes.

   .. example:: Defining a record

      The set of rational numbers may be defined as:

      .. coqtop:: reset all

         Record Rat : Set := mkRat
          { negative : bool
          ; top : nat
          ; bottom : nat
          ; Rat_bottom_nonzero : 0 <> bottom
          ; Rat_irreducible :
              forall x y z:nat, (x * y) = top /\ (x * z) = bottom -> x = 1
          }.

      The :n:`Rat_*` fields depend on :n:`top` and :n:`bottom`.
      :n:`Rat_bottom_nonzero` is a proof that :n:`bottom` (the denominator)
      is not zero.  :n:`Rat_irreducible` is a proof that the fraction is in
      lowest terms.

.. _reuse_field_name:

   .. example:: Reusing a field name in multiple records

      .. coqtop:: in

         Module A. Record R := { f : nat }. End A.
         Module B. Record S := { f : nat }. End B.

      .. coqtop:: all

         Check {| A.f := 0 |}.
         Check {| B.f := 0 |}.

.. _record_as_clause:

   .. example:: Using the "as" clause in a record definition

      .. coqtop:: all

         Record MyRecord := { myfield : nat } as VarName.
         About myfield. (* observe the MyRecord variable is named VarName *)
         

>          (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)

Reviewer 2: How does one rename the argument?

>          (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)

Reviewer 3: If I have
```coq
Definition foo (x : nat) : nat := ... .
```
But I want the name of x to be different, I use:
```coq
Arguments foo y : rename.
```
If you don't put rename coq will complain.

>          (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)

Reviewer 2: Hmm, if I use `Arguments bar HIMOM : rename`, what has to change in `Check (bar (HIMOM:={|bar:=0|}))`?

>          (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)

Reviewer 3: @jfehrle That will stay the same. You can see the current names of the variables using the `About` command. So if you look at `About bar.` you will see that it says `Arguments bar HIMOM`, so writing exactly that, even with a rename will do nothing. If we later found that we wanted to change HIMOM to HIDAD for whatever reason, we could achieve this by writing `Arguments bar HIDAD : rename.` Notably, the syntax `bar (HIMOM:=_)` wouldn't work anymore and we would instead have to write `bar (HIDAD:=_)`.

>          (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)

Reviewer 2: So with the "as HIMOM", you don't need ": rename" and vice-versa.  But both need an `Arguments` command.  Still a little puzzled why you use `bar (HIMOM:={|bar:=0|})` rather than just `bar {|bar:=0|}`.  Is it because the latter needs to be simplified to match the first?

>          (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)

Reviewer 3: The `f (x := _) (y := _)` syntax is very general, not just for projections. When working with functions/methods with many implicit arguments, it becomes impractical to use the `@` syntax for instance. For example, if I had `Arguments some_big_lemma {arg1 arg2 arg3} arg4`, it may always be that case that when arg4 is supplied, the implicit ones can be inferred by Coq. However say I am in a proof and I don't quite know what arg4 is going to be. I might therefore write `some_big_lemma _`. Now Coq ends up putting all these arguments as existential variables when checking it. But perhaps I want to be a bit more restrictive, how can I supply a value for arg2 for instance?

The usual answer is to use the `@` syntax. This makes all implicit arguments explicit. So now I can write `@some_big_lemma _ value_for_arg2 _ _`. However now imagine there are more than 20 arguments, you can easily see how this becomes impractical.

The point of the `some_big_lemma (arg2 := value_for_arg2) _` syntax is to make exactly this problem manageable.

         Arguments myfield {VarName}.   (* make "VarName" an implicit parameter *)
         Check myfield.
         Check (myfield (VarName:={| myfield := 0 |})).

   .. todo below: huh?  Hugo sez "the model to think about primitive projections

>       is not fully stabilized"

Reviewer 2: I meant to include it in the commit.

>       is not fully stabilized"

Reviewer: If it is committed, it should say something more precise than this. What is the point you are getting at?

            in order.  "mkRat" is defined by the Record command *)
         Definition half' := mkRat true 1 2 (O_S 1) one_two_irred.

Accessing fields (projections)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

   .. insertprodn term_projection term_projection

   .. prodn::
      term_projection ::= @term0 .( @qualid {? @univ_annot } {* @arg } )
      | @term0 .( @ @qualid {? @univ_annot } {* @term1 } )

   The value of a field can be accessed using *projection form* (:n:`@term_projection`,
   shown here) or with *application form* (see :n:`@term_application`) using the
   projection function associated with the field.  Don't forget the parentheses for the
   projection form.
   Glossing over some syntactic details, the two forms are:

   - :n:`@qualid__record.( {? @ } @qualid__field {* @arg })` (projection) and

   - :n:`{? @ } @qualid__field {* @arg } @qualid__record` (application)

   where the :n:`@arg`\s are the parameters of the inductive type.  If :n:`@` is
   specified, all implicit arguments must be provided.

   Since the projected object is part of the notation, it is always
   considered an explicit argument of :token:`qualid`, even if it is

>    formally declared as implicit (see :ref:`ImplicitArguments`).

Reviewer: I suppose this only applies in projection form. This could be worth making explicit.

Reviewer 2: This is almost unchanged from the original.  I don't understand "Since ... notation".
Would you care to suggest wording?

Reviewer: My suggestion is only to insert something like "In projection form," at the beginning.

>    .. exn:: @ident cannot be defined because it is informative and @ident is not

Reviewer: Clarifies that these are two different idents, also it looks like this is a warning, not an error:

```suggestion
   .. warn:: @ident__1 cannot be defined because it is informative and @ident__2 is not
```

This could also be:
```suggestion
   .. warn:: @ident__proj cannot be defined because it is informative and @ident__record is not
```

BTW, is this the right place to mention this warning, or should it be closer to the definition of record types?

      For example, :n:`Record foo:Prop := { x:Type }` generates the message
      "x cannot be defined ... and foo is not".  Proofs (objects of sort :n:`Prop`)
      are supposed to be non-distinguishable.  If you have two inhabitants of
      :n:`Type`, such as :n:`%{%| x := nat %|%}` and :n:`%{%| x := bool %|%}`, they are
      distinguishable (i.e. informative) and are therefore prohibited.

   .. example:: Accessing record fields

      .. coqtop:: all

         (* projection form *)
         Eval compute in half.(top).

      .. coqtop:: in

         Goal True.

      .. coqtop:: all

>          let x := eval compute in half.(top) in idtac x.

Reviewer: What is the point of repeating the same thing with both `Eval compute` and `eval compute`?

Reviewer 2: The second variation was not obvious to me.  Maybe it should be.  I'll drop it if you prefer.

         (* application form *)
         Eval compute in top half.
         Eval compute in bottom half.

>          Eval compute in Rat_bottom_nonzero half.

Reviewer: What is the point of showing three examples in application form but only one in projection form?

   .. example:: Matching on records

      .. coqtop:: all

         Eval compute in (
           match half with
           | {| negative := false; top := n |} => n
           | _ => 0
           end).

   .. example:: Accessing anonymous record fields with match

      .. coqtop:: in

         Record T := const { _ : nat }.
         Definition gett x := match x with const n => n end.
         Definition inst := const 3.

      .. coqtop:: all

         Eval compute in gett inst.

> The following settings let you control the display format for record types:

Reviewer: Apart from the `Printing Projection` flag, it seems to me that these flags and tables should be closer to the part on defining record types (and the note that was added there, maybe even moved to be part of the note).

Reviewer 2: I moved the notes and warnings into the record definition section.  The flags are all related to printing, so it seems reasonable to have them after the "Accessing" section.  Since it's short, I lean against making it a separate subsection, though I'll do that if you prefer.

Reviewer: I'm not so fond of having things at the end of a section that do not semantically belong to the last subsection...

.. flag:: Printing Records

   When this :term:`flag` is on (this is the default),
   use the record syntax (shown above) as the default display format.

You can override the display format for specified record types by adding entries to these tables:

.. table:: Printing Record @qualid

   This :term:`table` specifies a set of qualids which are displayed as records.  Use the
   :cmd:`Add` and :cmd:`Remove` commands to update the set of qualids.

.. table:: Printing Constructor @qualid

   This :term:`table` specifies a set of qualids which are displayed as constructors.  Use the
   :cmd:`Add` and :cmd:`Remove` commands to update the set of qualids.

.. flag:: Printing Projections

   Activates the projection form (dot notation) for printing projections (off by default).

   .. example::

      .. coqtop:: all

         Check top half.  (* off: application form *)
         Set Printing Projections.
         Check top half.  (* on:  projection form *)

> .. note:: Records exist in two flavors. In the first
>    implementation, a record :n:`@ident` with parameters :n:`{* @binder }`,
>    constructor :n:`@ident__0`, and fields :n:`{* @name @field_spec }`
>    is represented as a variant type with a single
>    constructor: :n:`Variant @ident {* @binder } : @sort := @ident__0
>    {* ( @name @field_spec ) }` and projections are defined by case analysis.
>    In the second implementation, records have
>    primitive projections: see :ref:`primitive_projections`.

Reviewer: This note is already present above so should be removed there (maybe a merge confict?):
```suggestion
```

> .. warn:: @ident cannot be defined.

Reviewer: This warning seems should be split into several and presented closer to `ident cannot be defined because it is informative and ident is not`. Furthermore, at least the first point is inaccurate (this would be an error, not a warning) and the reference to `Axiom` is a bit unexpected.

The next two error messages should also be moved.

Reviewer 2: I split off the first case.  It's not obvious how to map the other two cases to the messages generated by `Record.warning_or_error`, if indeed they match.  The third case may correspond to `MissingProj`, but I'm reluctant to describe it because I can't produce the message.  (We should have examples of how to produce every message even if we don't include all that in the doc.)

Reviewer: Food for thought: what about having an output test suite file dedicated to triggering all the documented error messages. Then we could have a mechanism to alert reviewers when a PR updates this test suite file without updating the documentation (@coqbot could post a comment for instance).

Reviewer 2: It would be good to have test suite files that generate as many of the messages as we can (some may be quite hard) just to document them for developers.  But how would we ensure that when a developer adds a new message that they remember to add it to the test suite?  Probably would need to invent a convention for tagging error message generation in the code so we can get a list of all error messages automatically.

Reviewer 2: PS Shouldn't it be lint instead of coqbot?

Reviewer 3: I believe it is possible to some extent already. Since we have errors in an open ended variant we could just print all the values. I'm sure people have done this before in OCaml. The issue remains that many errors in the codebase still rely on user_err etc. cc @ppedrot Who might be able to comment further.

Reviewer 4: It would be hard to do currently because we use the standard `exn` type. Even if we somehow managed to trick OCaml into printing the whole set of exn constructors (e.g. loading a byte plugin and relying on the REPL) we would get a lot of unrelated stuff in there. I'd rather have some kind of explicit registering mechanism, which we kind of have but is not expressive enough to give you access to the actual exception.

Reviewer: Folks: all your plans are fine, and I would definitely like some kind of registration mechanism for errors which can produce a dump of all known errors and can be checked against the documentation. But this is *not* what I was suggesting here. I was suggesting something more trivial (and therefore easier to implement): to adopt a *convention* that documented error messages are all listed in a specific test file, and to alert when this test file is modified but the documentation isn't (here I mentioned @coqbot rather than the linter because the alert would not be foolproof: it could happen that the test file needs to be updated but the doc doesn't, so it should not be blocking).

Reviewer 3: That could be done, but might not be so trivial. I guarantee there are errors in the code that have no known way to triggering. At least with exceptions we can trigger them all manually and see what they output. I think maybe we could do some grep magic on `user_err` and related things to list and perhaps later lint these errors. However even that has it's flaws.

Reviewer: >I guarantee there are errors in the code that have no known way to triggering.

If that's the case, shouldn't these errors be turned into anomalies?

In any case, my suggestion was not related to exhaustively listing all error messages in the documentation, but rather related to making sure that when error messages evolve or are removed, then the documentation of these error messages is kept up-to-date (we keep stumbling upon documented error messages that do not exist anymore). Do you also think that there are *documented* error messages that have no known way to triggering?

mathematical structures, such as groups or rings, hence the
synonym :cmd:`Structure`.

Defining record types
~~~~~~~~~~~~~~~~~~~~~

.. _record_grammar:

> .. cmd:: {| Record | Structure } @record_definition

Reviewer: I just noticed that `Record` actually support the `with` clause to simultaneously define several (non-recursive) records. @herbelin If we stop documenting this form, then we should also deprecate it.

Reviewer 2: It doesn't work.  IIRC @herbelin removed that recently (though when I look at the grammar, it looks like it should work-maybe a case where the parser can't fully handle the grammar as written).

![image](https://user-images.githubusercontent.com/1253341/154161650-3a2ffd53-ce4a-4438-b414-c39d81bb84c9.png)

Author: Yes, the `with` was removed in #15484.

Reviewer: Oh right, sorry I was not testing an up-to-date version.

   components, called :gdef:`fields <field>`, can be accessed by
   :gdef:`projection`. Such types can also be used to describe
   mathematical structures, such as groups or rings, hence the
   synonym :cmd:`Structure`.

   A :n:`@record_definition` defines a record named by
   :n:`@ident_decl`. The fields are given by the semicolon-separated
   list of :n:`@record_field` where :n:`@name` is the name of the
   field. Fields can be :gdef:`abstract <abstract field>` as in
   :n:`@name {* @binder } @of_type` or :gdef:`manifest <manifest
   field>`, that is explicitly specified, as in :n:`@name {* @binder } @of_type := @term`.
   Fields have a type given by :n:`@of_type` (when absent, as in :n:`@name {* @binder } := @term`,
   the type is inferred automatically).  In manifest
   fields, :n:`@term` is called the *body* of the field. The body
   and type of a field can be dependent from previous fields and the
   order of the fields is thus important.

   To a record is associated a constructor. The name of this
   constructor is provided by the occurrence of :n:`@ident` after
   :n:`:=`. When omitted, the constructor name defaults to

>    :n:`Build_@ident` where :n:`@ident` is the record name.

Reviewer 2: I tweaked the wording.  `ident_decl` is used many places in the grammar.  It's better not to expand it here (and everywhere else).

   :n:`@@qualid {?  @univ_annot } {+ @term1 }` where :n:`@qualid`
   refers to the constructor. That is, the constructor is seen as a
   function and the instances for the abstract fields are given in
   order as arguments of the constructor.

   In the record form, the
   instance of an abstract field is given as body of the
   corresponding field name and the instances can be given in any
   order. If a field is omitted, its instance is implicitly considered
   to be the :n:`_` placeholder. It can then be filled automatically
   by unification or by using obligations (see :ref:`programs`).

.. FIXME: move this to the main grammar in the spec chapter

Syntax of Record Projections
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

   .. insertprodn term_projection term_projection

   .. prodn::
      term_projection ::= @term0 .( @qualid {? @univ_annot } {* @arg } )
      | @term0 .( @ @qualid {? @univ_annot } {* @term1 } )

   The projection from an object can be written using either an
   *applicative syntax* of the form :n:`@term_application` (more
   precisely :n:`@qualid {?  @univ_annot } {+ @arg }` or
   :n:`@@qualid {?  @univ_annot } {+ @term1 }`) or a *projection

>    syntax* of the form :n:`@term_projection`. In both cases,

Reviewer 2: Maybe avoid repeating the definitions?

.. cmd:: {| Record | Structure } @record_definition
   :name: Record; Structure

   .. insertprodn record_definition field_type

   .. prodn::
      record_definition ::= {? > } @ident_decl {* @binder } {? : @sort } {? := {? @ident } %{ {*; @record_field } {? ; } %} {? as @ident } }
      record_field ::= {* #[ {*, @attribute } ] } @name {? @field_type } {? %| @natural }
      field_type ::= {* @binder } @of_type
      | {* @binder } @of_type := @term
      | {* @binder } := @term

   Defines a record type and creates projections for each field that has a name other
   than `_`. The field body and type can depend on previous
   fields, so the order of fields in the definition may matter.

>    :n:`{? > }`

Reviewer 2: Our convention is to describe items in the order they appear in the syntax so it's easy to find their descriptions rather than in order of importance.

>    :n:`{? > }`

Author: OK, if this is the convention.

   :n:`: @sort`
     The sort the record belongs to.  The default is :n:`Type`.

   :n:`:= {? @ident }`
     :n:`@ident` is the name of the record constructor.  If omitted,
     the name defaults to :n:`Build_@ident` where :n:`@ident` is the record name.

   :n:`as {? @ident}`
     Specifies the name used to refer to the argument corresponding to the
     record in the type of projections.

     .. todo PR: clarify above description, maybe give an example

   In :n:`@record_field`:

     :n:`@name` is the field name.  Since field names define projections, you can't
     reuse the same field name in two different records in the same scope.

     :n:`field_type` can be omitted only when the type of the field can be inferred
     from other fields. For example: the type of :n:`n` can be inferred from
     :n:`npos` in :n:`Record positive := { n; npos : 0 < n }`.

     :n:`| @natural`
       Specifies the priority of the field.  It is only allowed in :cmd:`Class` commands.

   In :n:`@field_type`:

     :n:`{+ @binder } : @of_type` is equivalent to :n:`forall {+ @binder } , @of_type`
     and similarly for the other forms of :n:`@field_type`.

>      .. todo PR: is the above correct for "binder := term"?

Reviewer 2: Is `:= forall binders, type := fun binders => term` correct?  Nesting the same binders seems odd/pointless.  Isn't it?

>      .. todo PR: is the above correct for "binder := term"?

Author: Was a typo. I meant `: forall binders, type := fun binders => term`.

> Nesting the same binders seems odd/pointless. Isn't it?

Yes, this is why there is a syntax to factorize them before the `:` and `:=`!

>      .. todo PR: is the above correct for "binder := term"?

Reviewer 2: OK, I used `:` for the equivalents of all three forms of `field_spec`.

      a record :n:`@ident` with parameters :n:`{* @binder }`,
      constructor :n:`@ident__0`, and fields :n:`{* @name @field_type }`
      is represented as a variant type with a single
      constructor: :n:`Variant @ident {* @binder } : @sort := @ident__0
      {* ( @name @field_type ) }` and projections are defined by case analysis.
      In the second implementation, records have
      primitive projections: see :ref:`primitive_projections`.

Constructing records
~~~~~~~~~~~~~~~~~~~~

   .. insertprodn term_record field_def

   .. prodn::
      term_record ::= %{%| {*; @field_def } {? ; } %|%}
      field_def ::= @qualid {* @binder } := @term

   Instances of record types can be constructed using either *record syntax*
   (:n:`@term_record`, shown here) or with *applicative syntax* (see :n:`@term_application`)
   using the constructor.

   .. todo PR: what does it mean if you provide a qualid in field_def that's not just an ident?
      Is that just to pick out a field defined in another section?

   In the record form, the fields can be given in any order.  Fields that can be
   inferred by unification or by using obligations (see :ref:`programs`) may be omitted.

>    .. todo PR: what is the difference between "mkRat ..." and "@mkRat ..."?

Reviewer 2: There was no description of `@`, so I added one.  But why isn't the second line in the grammar below `@ term1 {+ term1 }`?  Is that a bug or a feature?

![image](https://user-images.githubusercontent.com/1253341/153126449-9f11d4d4-7720-4f60-a3b0-7261496a5393.png)

>    .. todo PR: what is the difference between "mkRat ..." and "@mkRat ..."?

Author: >  Is that a bug or a feature?

Implicit arguments are attached to global names, so the prefix `@` applies only to global names.

This could somehow be generalized but that's another story... Currently, it applies only to global names (i.e. `@qualid`).

>    .. todo PR: what is the difference between "mkRat ..." and "@mkRat ..."?

Reviewer 2: The grammar here seems more complex than is necessary.  It would be easier to understand if we didn't have both `{+ arg }` and `{+ term1 }` in parallel productions.  I'd suggest simplifying `term_application` to just `{? @ } @term1 {+ @arg }`.

Given that `@` is a no-op for functions that don't have implicits, it shouldn't be a problem to replace `qualid_annotated` with `term1`.  For the first two lines of `arg`, Coq could give a runtime error pretty simply if the `term1` is not a `qualid_annotated`.

WDYT?

>    .. todo PR: what is the difference between "mkRat ..." and "@mkRat ..."?

Author: I think something like that was tried and it conflicted with uses of `@` as an infix symbol. The conclusion was something like that it would work by adopting a convention that for `@` to be used in the middle of a notation, it has either to be surrounded by two spaces or by no spaces at all, and when used to deactivate implicit arguments to be used with no space.

         Definition half := mkRat true 1 2 (O_S 1) one_two_irred.

Accessing fields (projections)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

   .. insertprodn term_projection term_projection

   .. prodn::
      term_projection ::= @term0 .( @qualid {? @univ_annot } {* @arg } )
      | @term0 .( @ @qualid {? @univ_annot } {* @term1 } )

   The value of a field can be accessed using *projection syntax* (:n:`@term_projection`,
   shown here) or with *applicative syntax* (see :n:`@term_application`) using the
   projection function associated with the field.
   Glossing over some syntactic details, the two forms are:

   - :n:`@qualid__record.(@qualid__field {* @arg })` (projective) and

   - :n:`@qualid__field {* @arg } @qualid__record` (applicative)

   where the :n:`@arg`\s are the parameters of the inductive type.

   Since the projected object is part of the notation, it is always
   considered an explicit argument of :token:`qualid`, even if it is
   formally declared as implicit (see :ref:`ImplicitArguments`).

   .. todo PR
      "Record foo2:Prop := { x:Type }." gives the output "foo2 is defined.
      x cannot be defined because it is informative and foo2 is not."

>       What does this mean?

Reviewer 2: I don't understand at all why they have to be non-distinguishable.  I see that `Record foo2:Prop := { x:Prop }.` doesn't work either.  Check what I added.  Not sure it's worth spending much time on.

>       What does this mean?

Author: > I don't understand at all why they have to be non-distinguishable.

That's the common intuition behind the notion of proposition compared to the notion of type. When observing a proposition, the idea is that only the provability matters and we don't want to be able to observe different proofs. Coq complies with this intuition (even if, syntactically, a proof is like a program).

>       What does this mean?

Reviewer 2: I get it that only provability matters.  I take it that's a choice made while implementing records.  Is the issue that Coq might try to compare 2 Props for equality, which would require additional computation, maybe a lot, to handle correctly?

It doesn't seem to make any difference using `SProp`.  Not sure that shouldn't make a difference.

   .. example::

      .. todo PR: maybe make this an example of matching?

      Let us define a function by pattern matching over a record:

      .. coqtop:: all

         Eval compute in (
           match half with
           | {| negative := false; top := n |} => n
           | _ => 0
           end).

   .. example:: Accessing record fields

      Let us project fields of a record, using either the applicative or projection syntax:

      .. coqtop:: all

         (* projection form *)
         Eval compute in half.(top).

         (* applicative form *)
         Eval compute in top half.
         Eval compute in bottom half.
         Eval compute in Rat_bottom_nonzero half.

The following settings let you control the display format for record types:

.. flag:: Printing Records

   When this :term:`flag` is on (this is the default),
   use the record syntax (shown above) as the default display format.

> 

Reviewer 2: I think we need a short description of primitive projections (what is it, what's is used for).  And it sounds like it's experimental (or at least will evolve), so we should say that.  Would you care to suggest some wording?

> 

Author: There is a section at the end of the chapter on primitive projections. I would say that until we reach a consensus on the form we want to give to primitive projections, it is not worth to change the current documentation.

> 

Author: Can we reasonably say that a feature used since almost 10 years that it is still experimental? Using "evolving" seems a good idea though, something like "note that the design of primitive projections is still liable to evolve"??

> 

Reviewer 2: 10 years seems a very long time to reach consensus!  I'll add a sentence about evolving.

> 

Author: This is a long time but it also reflects the time needed to understand the feature. At the very beginning it was motivated by an optimization of the representation of projections in memory. This interferred with implementing an emerging logical concept at this time called negative types, leading to two conceptually kinds of records (defined by a constructor, as in Variant, or by primitive destructors). Then, at some time, it was thought that one-constructor Variant could be emulated on top of primitive destructors, which was a mistake. Reverting this last step is not fully done. Etc.

   In :n:`@record_field`:

     :n:`@attribute`, if specified, can only be :n:`canonical`.

     :n:`@name` is the field name.  Since field names define projections, you can't
     reuse the same field name in two different records in the same scope.  This
     :ref:`example <reuse_field_name>` shows how to use the reuse the same field
     name in multiple records.

     :n:`@field_spec` can be omitted only when the type of the field can be inferred
     from other fields. For example: the type of :n:`n` can be inferred from
     :n:`npos` in :n:`Record positive := { n; npos : 0 < n }`.

     :n:`| @natural`
       Specifies the priority of the field.  It is only allowed in :cmd:`Class` commands.

   In :n:`@field_spec`:

     - :n:`{+ @binder } : @of_type` is equivalent to

>        :n:`forall {+ @binder } , @of_type`

Author: This should be ``:n:`: forall {+ @binder } , @of_type` `` (maybe originally my mistake).

     - :n:`{* @binder } := @term` is equivalent to

>        :n:`fun {* @binder } => @term`

Author: This should be ``:n:` := fun {* @binder } => @term` ``

Reviewer 2: With `:=` or (like the other two) `:`?

Also to be precise, I'll use `{+ binder }` in each of these:

![image](https://user-images.githubusercontent.com/1253341/153304381-5071c82f-d479-463a-a992-751bf1af3821.png)

versus `{* binder }` in the grammar:

![image](https://user-images.githubusercontent.com/1253341/153304468-a049aa9e-14f5-4ed1-a421-bdf26b25ccad.png)

Author: Indeed, this is only when there is at least one binder.

     - :n:`{* @binder } @of_type := @term` is equivalent to

>        :n:`forall {* @binder } , @type := fun {* @binder } => @term`

Author: Idem, should be `` :n:`: forall {* @binder } , @type := fun {* @binder } => @term`  ``

      .. coqtop:: all

>          Record MyRecord := { myfield : nat } as VarName.

Reviewer 2: Surely there is a less generic name than "VarName"?  How about "MyRecordType"?

Reviewer 3: I chose the generic name to really hammer home what it actually was. I feel like calling it `MyRecordType` would muddy the waters.

      .. coqtop:: all

         Record MyRecord := { myfield : nat } as VarName.
         About myfield. (* observe the MyRecord variable is named "VarName" *)
         
         (* make "VarName" implicit without having to rename the variable,
            which would be necessary without the "as" clause *)
         Arguments myfield {VarName}.   (* make "VarName" an implicit parameter *)
         Check myfield.
         Check (myfield (VarName:={| myfield := 0 |})).

   .. exn:: Records declared with the keyword Record or Structure cannot be recursive.

      The record name :token:`ident` appears in the type of its fields, but uses
      the :cmd:`Record` command. Use  the :cmd:`Inductive` or
      :cmd:`CoInductive` command instead.

   .. exn:: Cannot handle mutually (co)inductive records.

      Records cannot be defined as part of mutually inductive (or
      coinductive) definitions, whether with records only or mixed with

>       standard definitions.

Reviewer: It appears that this error message actually doesn't exist anymore because the first case ("with records only") is supported now. So the new error message to document instead is: "Mixed record-inductive definitions are not allowed".

Reviewer 2: I dropped this message.  The "Mixed ..." message is not primarily for Record, which doesn't allow constructors, so let's address it later.

Reviewer: Indeed, though it turns out that (currently) it can be raised by the `Record` command. But fine to keep this for the inductive chapter.

Reviewer 2: Since it sounds like Hugo will relax this restriction, likely dropping the message entirely, I'll do nothing on this now.

   .. exn:: @ident already exists

      The fieldname :n:`@ident` is already defined as a global.

   .. warn:: @ident cannot be defined.

     It can happen that the definition of a projection is impossible.
     This message is followed by an explanation of this impossibility.
     There may be two reasons:

     #. The :term:`body` of :token:`ident` uses an incorrect elimination for

>         :token:`ident` (see :cmd:`Fixpoint` and :ref:`Destructors`).

Reviewer 2: I have no idea.

     #. The type of the projections :token:`ident` depends on previous

>         projections which themselves could not be defined.

Reviewer 2: OK, but I don't understand the "incorrect elimination" message well enough to explain, nor do I know the text of that message.

>         projections which themselves could not be defined.

Reviewer 2: How do you know it's a warning?

>         projections which themselves could not be defined.

Reviewer: This is why I also asked the previous question. Obviously, we need someone else to answer (or to look at the code).

>    .. warn:: @ident__field cannot be defined because it is informative and @ident__record is not

Reviewer: I wonder whether it is really appropriate to use `field` to characterize the first ident: if I'm not mistaken, the field is correctly defined, it's just the projection that is not.

Reviewer 2: The grammar and text refers to the field identifier, not the projection identifier. 
 Shortly below, we appropriately refer to it as `ident__field`.  Not perfect but the best that can be done.  I added some words.

      For example, :n:`Record foo:Prop := { x:Type }` generates the message
      "x cannot be defined ... and foo is not".  Proofs (objects of sort :n:`Prop`)
      are supposed to be non-distinguishable.  If you have two inhabitants of
      :n:`Type`, such as :n:`%{%| x := nat %|%}` and :n:`%{%| x := bool %|%}`, they are
      distinguishable (i.e. informative) and are therefore prohibited.

   During the definition of the one-constructor inductive definition, all
   the errors of inductive definitions, as described in Section

>    :ref:`gallina-inductive-definitions`, may also occur.

Reviewer: This is impossible to understand without having first read the note that come afterward. Thus, I would at the very least move it after the note. But I also wonder how accurate it actually is...

Reviewer 2: Moved.

   Note here that the fields ``Rat_bottom_cond`` depends on the field ``bottom``
   and ``Rat_irred_cond`` depends on both ``top`` and ``bottom``.

.. example::

   Let us define the rational :math:`1/2` using either the applicative or record syntax:

    .. coqtop:: in

       Theorem one_two_irred : forall x y z:nat, x * y = 1 /\ x * z = 2 -> x = 1.
       Admitted.

       (* Applicative form: values for fields should be given in order *)
       Definition half := mkRat true 1 2 (O_S 1) one_two_irred.

       (* Record form: omitting inferrable fields top and bottom *)
       Definition half' :=
         {| sign := true;
            Rat_bottom_cond := O_S 1;
            Rat_irred_cond := one_two_irred |}.

.. example::

   Let us define a function by pattern matching over a record:

   .. coqtop:: all

      Eval compute in (
        match half with
        | {| sign := true; top := n |} => n
        | _ => 0
        end).

.. example::

   Let us project fields of a record, using either the applicative or projection syntax:

   .. coqtop:: all

      Eval compute in top half.
      Eval compute in bottom half.
      Eval compute in Rat_bottom_cond half.

>       Eval compute in half.(top).

Reviewer 2: Isn't there a simpler way to extract the value of a field that doesn't require `compute` or `Eval`?

This doesn't work--shouldn't it?

```
Goal True.
let x := half.top in idtac x.
```

Reviewer 2: Also I don't see how `Eval compute in top half.` produces `1`.  `top half` is 1 but I don't get what `compute` does with that.  Quite an opaque example.

   Note here that the fields ``Rat_bottom_cond`` depends on the field ``bottom``
   and ``Rat_irred_cond`` depends on both ``top`` and ``bottom``.

.. example::

>    Let us define the rational :math:`1/2` using either the applicative or record syntax:

Reviewer 2: Note that Coq comments in examples will appear in the manual--probably better to put "(\* applicative form \*)" and "(* record form *)" into the body of the example for readability.

Reviewer 2: The doc uses Python code to extract individual sentences and comments.  It's crude.  I think comments only work between sentences, not within them.

    .. coqtop:: in

       Theorem one_two_irred : forall x y z:nat, x * y = 1 /\ x * z = 2 -> x = 1.
       Admitted.

>        Definition half := mkRat true 1 2 (O_S 1) one_two_irred.

Reviewer 2: The definition of 'mkRat' should be shown or at least mention where it comes from (the user or from the `Record`).
The definition of the `sign` field should have a comment saying "true means positive".

       Definition half' :=
         {| sign := true;
            Rat_bottom_cond := O_S 1;

>             Rat_irred_cond := one_two_irred |}.

Reviewer 2: The definition seems more complex than necessary.  Perhaps it would be simpler to define `Rational` (rather than `Rat`) with only `top`, `bottom` and `sign` fields.

For `half'`, there should be some explanation about how `*_cond` binds `top` and `bottom`.  And maybe that should be a secondary example for maximum clarity.

         Admitted.

         (* Record form: top and bottom can be inferred from other fields *)
         Definition half :=
           {| negative := false;
              Rat_bottom_nonzero := O_S 1;
              Rat_irreducible := one_two_irred |}.

         (* Applicative form: use the constructor and provide values for all the fields

>             in order.  "mkRat" is the Record command *)

Author: You mean `Record constructor` I believe??

Reviewer 2: Typo, I left out "defined by".

      term_projection ::= @term0 .( @qualid {? @univ_annot } {* @arg } )
      | @term0 .( @ @qualid {? @univ_annot } {* @term1 } )

>    The value of a field can be accessed using *projective syntax* (:n:`@term_projection`,

Author: The word projective rings a bit like projective geometry to me. Would "projection" sound really bad?

Reviewer 2: I used "projection form" and "application form".    Personally, I like an adjective with a noun rather two nouns, but both "projective" and "projection" are OK.

Author: OK

   shown here) or with *applicative syntax* (see :n:`@term_application`) using the
   projection function associated with the field.  Don't forget the parentheses for the

>    projective form.

Author: "projection form"?

the errors of inductive definitions, as described in Section
:ref:`gallina-inductive-definitions`, may also occur.

> .. seealso:: Coercions and records in section :ref:`coercions-classes-as-records`.

Reviewer: What is this about? Should it be moved?

Reviewer 2: I moved it to the record definition section.  I didn't want to dig into the details of coercions right now (don't know much about them).

.. _primitive_projections:

Primitive Projections
~~~~~~~~~~~~~~~~~~~~~

Note: the design of primitive projections is still evolving.

When the :flag:`Primitive Projections` flag is on or the

> :attr:`projections(primitive)` attribute is supplied for a :n:`Record` definition, its

Reviewer: ```suggestion
:attr:`projections(primitive)` attribute is supplied for a :cmd:`Record` definition, its
```
